src/main/scala/is/hail/io/bgen/BgenWriter.scala

```diff
@@ -0,0 +1,298 @@
+    updateIntToBytesLE(bb, sampleBlockLength, 24)
+  }
+
+  def emitVariant(bb: ArrayBuilder[Byte], v: Variant, va: Annotation, gs: Iterable[Genotype],
+    rsidQuery: Option[Querier], varidQuery: Option[Querier], nSamples: Int, nBitsPerProb: Int) {
+    val nAlleles = v.nAlleles
+    require(nAlleles <= 0xffff, s"Maximum number of alleles per variant is ${ 0xffff }. Found ${ v.nAlleles }.")
+
+    val varid = varidQuery.flatMap(q => Option(q(va))).map(_.asInstanceOf[String]).getOrElse(v.toString)
+    val rsid = rsidQuery.flatMap(q => Option(q(va))).map(_.asInstanceOf[String]).getOrElse(".")
+
+    stringToBytesWShortLength(bb, varid)
+    stringToBytesWShortLength(bb, rsid)
+    stringToBytesWShortLength(bb, v.contig)
+    intToBytesLE(bb, v.start)
+    shortToBytesLE(bb, nAlleles)
+    stringToBytesWIntLength(bb, v.ref)
+    v.altAlleles.foreach(a => stringToBytesWIntLength(bb, a.alt))
+
+    val gtDataBlockStart = bb.length
+    intToBytesLE(bb, 0) // placeholder for length of compressed data
+    intToBytesLE(bb, 0) // placeholder for length of uncompressed data
+
+    val uncompressedData = emitGenotypeData(gs, nSamples, nAlleles, nBitsPerProb)
+
+    val uncompressedLength = uncompressedData.length
+    val compressedLength = compress(bb, uncompressedData)
+
+    updateIntToBytesLE(bb, compressedLength + 4, gtDataBlockStart)
+    updateIntToBytesLE(bb, uncompressedLength, gtDataBlockStart + 4)
+  }
+
+  def emitGenotypeData(gs: Iterable[Genotype], nSamples: Int, nAlleles: Int, nBitsPerProb: Int): Array[Byte] = {
+    val bb = new ArrayBuilder[Byte]
+
+    val nGenotypes = triangle(nAlleles)
+    val newSize = (1L << nBitsPerProb) - 1
+
+    intToBytesLE(bb, nSamples)
+    shortToBytesLE(bb, nAlleles)
+    bb += ploidy
+    bb += ploidy
+
+    val resized = new Array[Double](nGenotypes)
+    val index = new Array[Int](nGenotypes)
+    val fractional = new Array[Double](nGenotypes)
+
+    val samplePloidyStart = bb.length
+    var i = 0
+    while (i < nSamples) {
+      bb += 0x80.toByte // placeholder for sample ploidy - default is missing
+      i += 1
+    }
+
+    bb += phased
+    bb += nBitsPerProb.toByte
+
+    val bitPacker = new BitPacker(bb, nBitsPerProb)
+    gs.zipWithIndex.foreach { case (gt, idx) =>
+      val px = gt.unboxedPX
+      if (px == null) {
+        var i = 0
+        while (i < nGenotypes - 1) {
+          bitPacker += 0
+          i += 1
+        }
+      } else {
+        bb(samplePloidyStart + idx) = ploidy
+
+        val F = if (gt.isLinearScale) {
+          val conversionFactor = newSize.toDouble / 32768
+          resizeAndComputeFractional(px, resized, fractional, conversionFactor)
+        } else {
+          resizeAndComputeFractional(gt.unboxedGP, resized, fractional, newSize)
+        }
+
+        roundWithConstantSum(resized, fractional, index, bitPacker, F, newSize, keepLast = false)
+      }
+    }
+    bitPacker.flush()
+    bb.result()
+  }
+}
+
+trait IntConsumer {
```

Comment:
I added this abstract class because I didn't want to duplicate the `roundWithConstantSize` function, but I also wanted to be able to give the function both an Array[Byte] and an Array[Int] for the tests.